File: wpilibNewCommands/src/test/java/edu/wpi/first/wpilibj2/command/PrintCommandTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj2.command;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import org.junit.jupiter.api.Test;

class PrintCommandTest extends CommandTestBase {
  @Test
  void printCommandScheduleTest() {
    try (CommandScheduler scheduler = new CommandScheduler()) {
      final PrintStream originalOut = System.out;
      ByteArrayOutputStream testOut = new ByteArrayOutputStream();
      System.setOut(new PrintStream(testOut));
      try {
        PrintCommand command = new PrintCommand("Test!");

        scheduler.schedule(command);
        scheduler.run();

        assertFalse(scheduler.isScheduled(command));
        assertEquals(testOut.toString(), "Test!" + System.lineSeparator());
      } finally {
        System.setOut(originalOut);
      }
    }
  }
}

File: commandsv2/src/test/native/cpp/frc2/command/PrintCommandTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <frc2/command/Commands.h>

#include <regex>

#include "CommandTestBase.h"
#include "frc2/command/PrintCommand.h"

using namespace frc2;
class PrintCommandTest : public CommandTestBase {};

TEST_F(PrintCommandTest, PrintCommandSchedule) {
  CommandScheduler scheduler = GetScheduler();

  auto command = cmd::Print("Test!");

  testing::internal::CaptureStdout();

  scheduler.Schedule(command);
  scheduler.Run();

  EXPECT_TRUE(std::regex_search(testing::internal::GetCapturedStdout(),
                                std::regex("Test!")));

  EXPECT_FALSE(scheduler.IsScheduled(command));
}
